Integrations/Snowflake/Snowflake.py

```diff
@@ -0,0 +1,392 @@
+        The datetime/date/time/timedelta object to convert
+
+    returns:
+        Formatted string of the object
+    """
+    if isinstance(v, datetime):
+        return v.strftime('%Y-%m-%d %H:%M:%S.%f %z')
+    elif isinstance(v, date):
+        return v.strftime('%Y-%m-%d')
+    elif isinstance(v, dttime) or isinstance(v, timedelta):
+        return v.strftime('%H:%M:%S.%f')
+    return v
+
+
+def set_provided(params, key, val1, val2=None):
+    """
+    If value is provided, set it in the dict
+    """
+    if val1:
+        params[key] = val1
+    elif val2:
+        params[key] = val2
+
+
+def process_table_row(row, checks):
+    """
+    Check row data and reformat if necessary
+
+    parameter: (dict) row
+        The data (table row) that needs to be processed
+
+    parameter: (dict[str, list]) checks
+        Dictionary where the key is a string indicative of the type (or bucket of types) that needs
+        reformatting and the values are a list of column names whose data is of that type
+
+    returns:
+        Reformatted Row
+    """
+    for column_name, val in row.items():
+        if column_name in checks.get('isDecimal', []):
+            # Then check the value and reformat it if necessary
+            if type(val) == Decimal:
+                row[column_name] = str(val)
+        elif column_name in checks.get('isDT', []):
+            # Then reformat it if necessary
+            row[column_name] = convert_datetime_to_string(val)
+    return row
+
+
+def format_to_json_serializable(column_descriptions, results):
+    """
+    Screen and reformat any data in 'results' argument that is
+    not json serializable, and return 'results'. 'results' can
+    be a table of data (a list of rows) or a single row.
+
+    parameter: (list) column_descriptions
+        The metadata that describes data for each column in the 'results' parameter
+
+    parameter: (list/dict) results
+        What was returned by the cursor object's execute or fetch operation
+
+    returns:
+        Reformatted 'results'
+    """
+    name = 0
+    type_code = 1
+
+    checks = {}
+    # Screen by type_code
+    for col in column_descriptions:
+        if col[type_code] == 0:
+            # Then need to check that column's data to see if its data type is Decimal
+            checks.setdefault('isDecimal', []).append(col[name])
+        elif col[type_code] in {3, 4, 6, 7, 8, 12}:
+            # Then need to check that column's data to see if its data type is date, time, timedelta or datetime
+            checks.setdefault('isDT', []).append(col[name])
+
+    # if 'results' is a list then it is a data table (list of rows) and need to process each row
+    # in the table, otherwise if 'results' is a dict then it a single table row
+    # Check candidates and reformat if necessary
+    if isinstance(results, dict):
+        results = process_table_row(results, checks)
+    else:
+        # if 'results' isn't a dict, assume it's a list
```

Comment:
why can we assume that?

Reply:
A query returns either a table ( a list of rows where each row is a dict ) or a row ( which is a dict ) even in the case where one cell is being returned it's still returned as a dict.